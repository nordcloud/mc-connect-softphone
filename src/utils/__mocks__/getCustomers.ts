export const getCustomers = jest.fn().mockResolvedValue({
  CustomerId: 'dummyCustomerId',
  CustomerName: 'dummyCustomerName',
  OutboundNumber: 'dummyOutboundNumber',
});
