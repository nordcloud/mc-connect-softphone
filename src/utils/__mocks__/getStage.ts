export const getStage = jest.fn().mockReturnValue('prod');
