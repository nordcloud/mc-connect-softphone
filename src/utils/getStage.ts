import { Stage } from '../types';
import { assertStageName } from './assertStageName';

export function getStage(): Stage {
  const { STAGE } = process.env;

  assertStageName(STAGE);

  return STAGE;
}
