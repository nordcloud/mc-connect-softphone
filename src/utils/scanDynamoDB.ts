import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import {
  DynamoDBDocumentClient,
  ScanCommand,
  ScanCommandInput,
} from '@aws-sdk/lib-dynamodb';

const ddbDocClient = DynamoDBDocumentClient.from(new DynamoDBClient({}));

type Props<Item> = {
  tableName: string;
  onlyAttributes?: (keyof Item)[];
};

export async function scanDynamoDB<Item>({ tableName, onlyAttributes }: Props<Item>) {
  let results: Item[] = [];

  const scan = async (ExclusiveStartKey?: Record<string, unknown>) => {
    const input: ScanCommandInput = {
      TableName: tableName,
      ExclusiveStartKey,
    };

    if (onlyAttributes) {
      input.ProjectionExpression = onlyAttributes.join(',');
    }

    const command = new ScanCommand(input);
    const { Items, LastEvaluatedKey } = await ddbDocClient.send(command);

    if (Items) {
      results = results.concat(Items as Item[]);
    }

    if (LastEvaluatedKey) {
      await scan(LastEvaluatedKey);
    }
  };

  await scan();

  return results;
}
