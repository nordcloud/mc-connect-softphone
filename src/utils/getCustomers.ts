import { CUSTOMERS_TABLE_ARN } from '../consts';
import { customersMock } from '../mockData';
import { CustomersTableItem } from '../types';
import { getStage } from './getStage';
import { scanDynamoDB } from './scanDynamoDB';

export async function getCustomers() {
  if (getStage() === 'local') {
    return customersMock;
  }

  const tableName = CUSTOMERS_TABLE_ARN.split('/').at(-1) as string;
  const items = await scanDynamoDB<CustomersTableItem>({ tableName });
  const sortedItems = items.sort((a, b) => a.CustomerName.localeCompare(b.CustomerName));

  return sortedItems;
}
