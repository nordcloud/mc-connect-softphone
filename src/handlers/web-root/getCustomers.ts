import { APIGatewayProxyEventV2 } from 'aws-lambda';

import { CUSTOMERS_TABLE_ARN } from '../../consts';
import { customersMock } from '../../mockData';
import { CustomersTableItem } from '../../types';
import { getStage } from '../../utils/getStage';
import { scanDynamoDB } from '../../utils/scanDynamoDB';

export async function getCustomers(event: APIGatewayProxyEventV2) {
  const isMockMode =
    getStage() === 'local' &&
    Object.keys(Object(event.queryStringParameters)).includes('mock_db');

  if (isMockMode) {
    return customersMock;
  }

  const tableName = CUSTOMERS_TABLE_ARN.split('/').at(-1);

  if (!tableName) {
    throw new Error('Missing customers table name');
  }

  const items = await scanDynamoDB<CustomersTableItem>({ tableName });
  const sortedItems = items.sort((a, b) => a.CustomerName.localeCompare(b.CustomerName));

  return sortedItems;
}
