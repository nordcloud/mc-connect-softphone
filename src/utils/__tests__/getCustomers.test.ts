import { APIGatewayProxyEventV2 } from 'aws-lambda';

import { getCustomers } from '../getCustomers';
import { getStage } from '../getStage';
import { scanDynamoDB } from '../scanDynamoDB';

jest.mock('../getStage');

jest.mock('../scanDynamoDB', () => ({
  scanDynamoDB: jest
    .fn()
    .mockResolvedValue([{ CustomerName: 'xyz' }, { CustomerName: 'abc' }]),
}));

const event = {} as APIGatewayProxyEventV2;

describe('getCustomers', () => {
  it('calls scanDynamoDB with correct args', async () => {
    await getCustomers(event);
    expect(jest.mocked(scanDynamoDB).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    expect(getCustomers(event)).resolves.toMatchSnapshot();
  });

  describe('when running in mock mode', () => {
    const eventClone = {
      queryStringParameters: { mock_db: '' },
    } as unknown as APIGatewayProxyEventV2;

    it('resolves with a correct value in "prod" stage', () => {
      expect(getCustomers(eventClone)).resolves.toMatchSnapshot();
    });

    it('resolves with a correct value in "local" stage', () => {
      jest.mocked(getStage).mockReturnValueOnce('local');
      expect(getCustomers(eventClone)).resolves.toMatchSnapshot();
    });
  });
});
