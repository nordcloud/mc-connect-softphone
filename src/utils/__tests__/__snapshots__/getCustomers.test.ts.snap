// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCustomers calls scanDynamoDB with correct args 1`] = `
[
  [
    {
      "tableName": "Connect-Outbound-Calling-Table",
    },
  ],
]
`;

exports[`getCustomers resolves with a correct value 1`] = `
[
  {
    "CustomerName": "abc",
  },
  {
    "CustomerName": "xyz",
  },
]
`;

exports[`getCustomers when running in mock mode resolves with a correct value in "local" stage 1`] = `
[
  {
    "CustomerId": "mockedCustomerId",
    "CustomerName": "mockedCustomerName",
    "OutboundNumber": "mockedCustomerOutboundNumber",
  },
]
`;

exports[`getCustomers when running in mock mode resolves with a correct value in "prod" stage 1`] = `
[
  {
    "CustomerName": "abc",
  },
  {
    "CustomerName": "xyz",
  },
]
`;
